### DISCUSSION TITLE
Storage API, uploading Blobs and Array Buffers

### DISCUSSION THREAD

**Discussion by leo-petrucci** [supabase/supabase#2047] [Questions] [answered]
I'm in need of uploading a blob or an Array Buffer to Supabase Storage. This is being done server-side, so essentially I have no way to create a `File` to upload.

I came across [this PR that adds support for uploading array buffers](https://github.com/supabase/storage-api/pull/29), but if I try to use it I just get errors back from the API:

```
        // the actual image is fetched from an url, this is just an example
        const { size } = fs.statSync(process.cwd() + '/image.png');
        const buffer2 = fs.readFileSync(process.cwd() + '/image.png');

        axios({
          method: 'POST',
          url: `${SUPABASE_URL}/storage/v1/object/avatars/pic.png`,
          headers: {
            'Content-Type': 'image/png',
            'Content-Length': size,
            Authorization: `Bearer ${SUPABASE_KEY}`,
          },
          data: buffer2,
        })
          .then((res) => console.log(res))
          .catch((res) => console.log(res));
```

I get this response:

```
    data: {
      statusCode: '42501',
      error: '',
      message: 'new row violates row-level security policy for table "objects"'
    }
```

What am I doing wrong?

**Comment by inian** [comment, marked as answer]
Hi @creativiii, 

If you are doing this from the server side, you can try this with the service key instead of the anon key which will bypass Row Level Security.

**Reply by leo-petrucci** [reply]
That's exactly what I was missing! Thank you 🥳

**Reply by leopsidom** [reply]
What if the error happens from client side ?

**Reply by leopsidom** [reply]
Never mind. I resolved it by creating policies under the storage bucket to allow insert operation for anon users. 

Btw, really love the great product !!
